[FIX JENKINS-40700] Display communication protocol in agent logs

Aki-7 opened the pull request:
I'm working on the ticket below for my first contribution to Jenkins!
[[JENKINS-40700] Display communication  protocol in agent logs](https://issues.jenkins.io/browse/JENKINS-40700)

As it is difficult to know the communication protocol from the established channel on the controller side, I propose holding the communication protocol name on the agent side. It is much simpler than holding it on the controller side.  

So, I submitted the PR to [jenkinsci / remoting](https://github.com/jenkinsci/remoting) that enables remote agents to hold the communication protocol name.
link: https://github.com/jenkinsci/remoting/pull/450
link: https://github.com/jenkinsci/remoting/pull/452

And what I do here is
- Fetch the communication protocol name from the agent
- Display it in the agent logs
- Display the Launcher class name.
  - Only displaying the communication protocol name can be misleading for users.
    (see SSH Build Agents Plugin log example below) 

The agent logs will be like this.

Launch with default JNLPLauncher:
```
Remoting version: 4.8
Launcher: JNLPLauncher
Communication Protocol: JNLP4-connect
This is a Unix agent
Agent successfully connected and online
```

Launch with default JNLPLauncher using WebSocket:
```
Remoting version: 4.8
Launcher: JNLPLauncher
Communication Protocol: WebSocket
This is a Unix agent
Agent successfully connected and online
```

Launch with SSH Build Agents Plugin:
```
Remoting version: 4.8
Launcher: SSHLauncher
Communication Protocol: Standard in/out
This is a Unix agent
Evacuated stdout
Agent successfully connected and online
```

### Proposed changelog entries

* Display communication protocol in agent logs.

<!-- Comment: 
The changelogs will be integrated by the core maintainers after the merge.  See the changelog examples here: https://jenkins.io/changelog/ -->

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [x] (If applicable) Jira issue is well described
- [x] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [x] Appropriate autotests or explanation to why this change has no tests
  - No tests because this is a small change.
- [x] For dependency updates: links to external changelogs and, if possible, full diffs

<!-- For new API and extension points: Link to the reference implementation in open-source (or example in Javadoc) -->

### Desired reviewers

@oleg-nenashev 

<!-- Comment:
If you need an accelerated review process by the community (e.g., for critical bugs), mention @jenkinsci/code-reviewers
-->

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are correct
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins-ci.org/issues/?filter=12146)).

Aki-7 commented on core/src/main/java/hudson/slaves/SlaveComputer.java:
```diff
@@ -992,6 +998,18 @@ public String call() throws IOException {
+            try {
+                Engine engine = Engine.current();
+                Method getProtocolName = Engine.class.getMethod("getProtocolName");
+                return (String) getProtocolName.invoke(engine);
```
These three lines are same as
```java
return Engine.current().getProtocolName();
```
but `Engine.current().getProtocolName()` will be a compile error because `remoting.Engine` doesn't have `getProtocolName` method in the remoting version pointed in pom.xml.
So, after merging the PR in the remoting module, releasing remoting:4.8 and dump remoting version pointed in pom.xml to 4.8, We can change these lines to
```java
return Engine.current().getProtocolName();
```
I'm waiting for the feedback on this situation.

res0nance replied:
Please add a TODO to detail this so the future reader will correct it when the time comes.

res0nance reviewed (commented):
A test to ensure this functionality works would be nice.

This PR also might want to bump the version of remoting to the incremental for testing and update it together with this bit.

Aki-7 commented:
> This PR also might want to bump the version of remoting to the incremental for testing and update it together with this bit.

Yes, but in my understanding, to test this feature, we need to release remoting-4.8 first including https://github.com/jenkinsci/remoting/pull/450. So, if the changes in https://github.com/jenkinsci/remoting/pull/450 look good, how about putting this PR on hold until the next release of remoting?

res0nance commented:
> > This PR also might want to bump the version of remoting to the incremental for testing and update it together with this bit.
> 
> Yes, but in my understanding, to test this feature, we need to release remoting-4.8 first including [jenkinsci/remoting#450](https://github.com/jenkinsci/remoting/pull/450). So, if the changes in [jenkinsci/remoting#450](https://github.com/jenkinsci/remoting/pull/450) look good, how about putting this PR on hold until the next release of remoting?

You can technically use the incremental, see https://repo.jenkins-ci.org/incrementals/org/jenkins-ci/main/remoting/4.8-rc2918.d1f5e04493a9/ to continue working on the tests. Then once its done we can wait till remoting is released before updating it to the official 4.8 remoting.

jeffret-b commented:
Yes, please update the Remoting version in the pom to reference the incremental version and then finish up with this to use the mechanisms properly.

Aki-7 commented:
During writing the test and understanding the structure of remoting modules more, I came up with a better log style and I'll change this PR a little.

The agent logs will be like this.

Launch with default JNLPLauncher:
```
Remoting version: 4.8
Launcher: JNLPLauncher
Communication Protocol: JNLP4-connect
This is a Unix agent
Agent successfully connected and online
```

Launch with default JNLPLauncher using WebSocket:
```
Remoting version: 4.8
Launcher: JNLPLauncher
Communication Protocol: WebSocket
This is a Unix agent
Agent successfully connected and online
```

Launch with SSH Build Agents Plugin:
```
Remoting version: 4.8
Launcher: SSHLauncher
Communication Protocol: Standard in/out
This is a Unix agent
Evacuated stdout
Agent successfully connected and online
```

I submitted additional PR to remoting for this feature.
link: https://github.com/jenkinsci/remoting/pull/452

I'll push the extra commit to this branch after the incremental is created.

oleg-nenashev commented on core/src/main/java/hudson/slaves/SlaveComputer.java:
```diff
@@ -992,6 +999,18 @@ public String call() throws IOException {
+        @Override
+        public String call() throws IOException {
+            try {
+                if (Engine.current() != null)
```
Static analysis will be unhappy unless you assign the value to a local variabe

Aki-7 replied:
Oh, Thank you! I'll fix it.

oleg-nenashev commented on pom.xml:
```diff
@@ -99,7 +99,7 @@ THE SOFTWARE.
 
     <!-- Bundled Remoting version -->
-    <remoting.version>4.7</remoting.version>
+    <remoting.version>4.8-rc2925.e195d94d1423</remoting.version>
```
We will need to list all notable changes from https://github.com/jenkinsci/remoting/releases/tag/remoting-4.8 in the changelog proposal, not just your change

Aki-7 replied:
I got it.

Aki-7 replied:
I'll change the changelog proposal and PR title after the new remoting release including https://github.com/jenkinsci/remoting/pull/452. Sorry for the bad timing additional PR...

oleg-nenashev commented on core/src/main/java/hudson/slaves/SlaveComputer.java:
```diff
@@ -992,6 +999,18 @@ public String call() throws IOException {
+                if (Engine.current() != null)
+                    return Engine.current().getProtocolName();
+                return Launcher.getCommunicationProtocolName();
+            } catch (Throwable ex) { return null; } // remote not support this feature
```
This code will catch `Error`s which are expected to be fatal in the most of cases. Is it intended?

Aki-7 replied:
I referred to the other part of the code, but not good at least for this change. I'll fix it.

oleg-nenashev commented on core/src/main/java/hudson/slaves/SlaveComputer.java:
```diff
@@ -651,6 +652,12 @@ public void onClosed(Channel c, IOException cause) {
+        log.println("Launcher: " + getLauncher().getClass().getSimpleName());
+
+        String communicationProtocol = channel.call(new CommunicationProtocol());
+        if (communicationProtocol != null)
+            log.println("Communication Protocol: " + communicationProtocol);
+
```
```suggestion
        if (communicationProtocol != null) {
            log.println("Communication Protocol: " + communicationProtocol);
        }

```

Aki-7 committed:
[FIX JENKINS-40700] Display Laucher and protocol in agent logs.

Fetch the communication protocol from agent, get the Launcher class
name and display them in agent logs.

oleg-nenashev committed:
Merge branch 'master' into agent-protocol-logs

oleg-nenashev reviewed (approved):
Looks good to me. Remoting 4.8 was merged in another pull request, but we still need to integrate this logic for a full fix

Aki-7 commented:
Thank you @oleg-nenashev!  But as the commits of https://github.com/jenkinsci/remoting/pull/452 are not included in Remoting 4.8, we have to put this PR on hold until the next Remoting release. Is this correct?

timja commented:
> Thank you @oleg-nenashev! But, because the commits of [jenkinsci/remoting#452](https://github.com/jenkinsci/remoting/pull/452) are not included in Remoting 4.8, we have to put this PR on hold until the next Remoting release. Is this correct?

we're now using 4.10, is there anything left on this?

alecharp commented:
@Aki-7 could you please attend to the merge conflicts here? This way we can move this PR forward as it already has more approval than needed to be merged.

alecharp commented:
@Aki-7 for the moment, I'm marking this PR as `stalled`, this means that, in case you are not able to attend to it, anyone in the community can try to finish it. Of course, that doesn't exclude you. In case no one is able to complete your initial work, we will have to close this PR in the future.

basil committed:
Merge remote-tracking branch 'origin/master' into agent-protocol-logs

basil committed:
Simplify test

basil commented:
This PR is now ready for merge. We will merge it after approximately 24 hours if there is no negative feedback. Please see the [merge process documentation](https://github.com/jenkinsci/jenkins/blob/master/docs/MAINTAINERS.adoc#merge-process) for more information about the merge process. Thanks!

basil commented on test/src/test/java/hudson/slaves/SlaveComputerTest.java:
```diff
@@ -58,6 +63,17 @@
+        Assert.assertTrue(log.contains("Remoting version: " + Launcher.VERSION));
+        Assert.assertTrue(log.contains("Launcher: " + SimpleCommandLauncher.class.getSimpleName()));
+        Assert.assertTrue(log.contains("Communication Protocol: Standard in/out"));
+        Assert.assertTrue(log.contains("This is a Unix agent"));
```
This will probably fail on our Windows test agents though

basil committed:
Fix test on Windows?
